From 6437bca6ed96c59ecafc53f0dde989eeaf8d69b8 Mon Sep 17 00:00:00 2001
From: Frank Ramsay <framsay@redhat.com>
Date: Thu, 9 Oct 2014 19:37:20 -0400
Subject: [mm] thp: add VM_INIT_DEF_MASK and PRCTL_THP_DISABLE

Message-id: <1412883441-28028-4-git-send-email-framsay@redhat.com>
Patchwork-id: 97306
O-Subject: [RHEL7.1 PATCH 3/4] mm, thp: add VM_INIT_DEF_MASK and PRCTL_THP_DISABLE
Bugzilla: 1149786
RH-Acked-by: Oleg Nesterov <oleg@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Rik van Riel <riel@redhat.com>

Bugzilla: 1149786
Upstream Status: upstream commit: a0715cc22601e8830ace98366c0c2bd8da52af52
Build Info: 8080538
Tested: Tested on SGI UV system.

Commit a0715cc22601e8830ace98366c0c2bd8da52af52
Author: Alex Thorlton <athorlton@sgi.com>
Date: Mon, 7 Apr 2014 15:37:10 -0700

(A minor edit to this patch was required as the upstream
mm.h is enough ahead of the current RHEL tree to
prevent applying the patch.)

Add VM_INIT_DEF_MASK, to allow us to set the default flags for VMs.  It
also adds a prctl control which allows us to set the THP disable bit in
mm->def_flags so that VMs will pick up the setting as they are created.

 Signed-off-by: Alex Thorlton <athorlton@sgi.com>
 Suggested-by: Oleg Nesterov <oleg@redhat.com>
 Cc: Gerald Schaefer <gerald.schaefer@de.ibm.com>
 Cc: Martin Schwidefsky <schwidefsky@de.ibm.com>
 Cc: Heiko Carstens <heiko.carstens@de.ibm.com>
 Cc: Christian Borntraeger <borntraeger@de.ibm.com>
 Cc: Paolo Bonzini <pbonzini@redhat.com>
 Cc: "Kirill A. Shutemov" <kirill.shutemov@linux.intel.com>
 Cc: Mel Gorman <mgorman@suse.de>
 Acked-by: Rik van Riel <riel@redhat.com>
 Cc: Ingo Molnar <mingo@kernel.org>
 Cc: Peter Zijlstra <peterz@infradead.org>
 Cc: Andrea Arcangeli <aarcange@redhat.com>
 Cc: Oleg Nesterov <oleg@redhat.com>
 Cc: "Eric W. Biederman" <ebiederm@xmission.com>
 Cc: Alexander Viro <viro@zeniv.linux.org.uk>
 Cc: Johannes Weiner <hannes@cmpxchg.org>
 Cc: David Rientjes <rientjes@google.com>
 Cc: Paolo Bonzini <pbonzini@redhat.com>
 Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
 Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/mm.h b/include/linux/mm.h
index e1a8eec..e341f6d 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -171,6 +171,9 @@ extern unsigned int kobjsize(const void *objp);
  */
 #define VM_SPECIAL (VM_IO | VM_DONTEXPAND | VM_PFNMAP)
 
+/* This mask defines which mm->def_flags a process can inherit its parent */
+#define VM_INIT_DEF_MASK VM_NOHUGEPAGE
+
 /*
  * mapping from the currently active vm_flags protection bits (the
  * low four bits) to a page protection mask..
diff --git a/include/uapi/linux/prctl.h b/include/uapi/linux/prctl.h
index 289760f..58afc04 100644
--- a/include/uapi/linux/prctl.h
+++ b/include/uapi/linux/prctl.h
@@ -149,4 +149,7 @@
 
 #define PR_GET_TID_ADDRESS 40
 
+#define PR_SET_THP_DISABLE 41
+#define PR_GET_THP_DISABLE 42
+
 #endif /* _LINUX_PRCTL_H */
diff --git a/kernel/fork.c b/kernel/fork.c
index 6035582..85f97e8 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -534,8 +534,6 @@ static struct mm_struct *mm_init(struct mm_struct *mm, struct task_struct *p)
  atomic_set(&mm->mm_count, 1);
  init_rwsem(&mm->mmap_sem);
  INIT_LIST_HEAD(&mm->mmlist);
- mm->flags = (current->mm) ?
-  (current->mm->flags & MMF_INIT_MASK) : default_dump_filter;
  mm->core_state = NULL;
  atomic_long_set(&mm->nr_ptes, 0);
  memset(&mm->rss_stat, 0, sizeof(mm->rss_stat));
@@ -546,8 +544,15 @@ static struct mm_struct *mm_init(struct mm_struct *mm, struct task_struct *p)
  mm_init_owner(mm, p);
  clear_tlb_flush_pending(mm);
 
- if (likely(!mm_alloc_pgd(mm))) {
+ if (current->mm) {
+  mm->flags = current->mm->flags & MMF_INIT_MASK;
+  mm->def_flags = current->mm->def_flags & VM_INIT_DEF_MASK;
+ } else {
+  mm->flags = default_dump_filter;
   mm->def_flags = 0;
+ }
+
+ if (likely(!mm_alloc_pgd(mm))) {
   mmu_notifier_mm_init(mm);
   return mm;
  }
diff --git a/kernel/sys.c b/kernel/sys.c
index 2e7076b..731fb8e 100644
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -2260,6 +2260,21 @@ SYSCALL_DEFINE5(prctl, int, option, unsigned long, arg2, unsigned long, arg3,
   if (arg2 || arg3 || arg4 || arg5)
    return -EINVAL;
   return current->no_new_privs ? 1 : 0;
+ case PR_GET_THP_DISABLE:
+  if (arg2 || arg3 || arg4 || arg5)
+   return -EINVAL;
+  error = !!(me->mm->def_flags & VM_NOHUGEPAGE);
+  break;
+ case PR_SET_THP_DISABLE:
+  if (arg3 || arg4 || arg5)
+   return -EINVAL;
+  down_write(&me->mm->mmap_sem);
+  if (arg2)
+   me->mm->def_flags |= VM_NOHUGEPAGE;
+  else
+   me->mm->def_flags &= ~VM_NOHUGEPAGE;
+  up_write(&me->mm->mmap_sem);
+  break;
  default:
   error = -EINVAL;
   break;
-- 
1.7.1